File: frontend/src/api/detection.ts
import api from './index'
import type { Detection, PaginatedResponse, AnalysisResult, PropagationNode } from '@/types'

export interface DetectionRequest {
  content: string
  include_analysis?: boolean
  include_propagation?: boolean
}

export interface BatchDetectionRequest {
  contents: string[]
  include_analysis?: boolean
}

export interface BatchDetectionResponse {
  total: number
  success: number
  failed: number
  results: Detection[]
}

export interface HistoryFilters {
  page?: number
  page_size?: number
  is_rumor?: boolean
  risk_level?: string
  start_date?: string
  end_date?: string
}

export interface PropagationResponse {
  detection_id: string
  nodes: PropagationNode[]
  pattern?: string
  spread_speed?: string
  estimated_reach?: number
  influence_score?: number
}

export const detectionApi = {
  async detectSingle(data: DetectionRequest): Promise<Detection> {
    const response = await api.post<Detection>('/detection/single', data)
    return response.data
  },

  async detectBatch(data: BatchDetectionRequest): Promise<BatchDetectionResponse> {
    const response = await api.post<BatchDetectionResponse>('/detection/batch', data)
    return response.data
  },

  async getDetection(id: string): Promise<Detection> {
    const response = await api.get<Detection>(`/detection/${id}`)
    return response.data
  },

  async getAnalysis(id: string): Promise<AnalysisResult> {
    const response = await api.get<AnalysisResult>(`/detection/${id}/analysis`)
    return response.data
  },

  async getPropagation(id: string): Promise<PropagationResponse> {
    const response = await api.get<PropagationResponse>(`/detection/${id}/propagation`)
    return response.data
  },

  async getHistory(filters: HistoryFilters = {}): Promise<PaginatedResponse<Detection>> {
    const params = new URLSearchParams()
    if (filters.page) params.append('page', filters.page.toString())
    if (filters.page_size) params.append('page_size', filters.page_size.toString())
    if (filters.is_rumor !== undefined) params.append('is_rumor', filters.is_rumor.toString())
    if (filters.risk_level) params.append('risk_level', filters.risk_level)
    if (filters.start_date) params.append('start_date', filters.start_date)
    if (filters.end_date) params.append('end_date', filters.end_date)

    const response = await api.get<PaginatedResponse<Detection>>(`/history?${params.toString()}`)
    return response.data
  },

  async deleteDetection(id: string): Promise<void> {
    await api.delete(`/history/${id}`)
  },

  async deleteMultiple(ids: string[]): Promise<void> {
    await api.delete('/history/batch', { data: { ids } })
  },
}

File: backend/app/schemas/detection.py
"""Detection schemas for request/response validation."""

import uuid
from datetime import datetime
from enum import Enum
from typing import Optional

from pydantic import BaseModel, Field


class RiskLevel(str, Enum):
    """Risk level enumeration for detection results."""

    LOW = "low"
    MEDIUM = "medium"
    HIGH = "high"
    CRITICAL = "critical"

    @classmethod
    def from_confidence(cls, confidence: float, is_rumor: bool) -> "RiskLevel":
        """Determine risk level based on confidence and rumor status."""
        if not is_rumor:
            return cls.LOW

        # For rumors: lower confidence in being true = higher risk
        if confidence >= 0.8:
            return cls.LOW
        elif confidence >= 0.6:
            return cls.MEDIUM
        elif confidence >= 0.4:
            return cls.HIGH
        else:
            return cls.CRITICAL


class AnalysisResult(BaseModel):
    """Schema for detailed analysis result."""

    keywords: list[str] = Field(default_factory=list)
    sentiment: str = "neutral"
    category: str = "other"
    sources: list[str] = Field(default_factory=list)
    fact_check_points: list[str] = Field(default_factory=list)
    risk_indicators: list[str] = Field(default_factory=list)


class DetectionRequest(BaseModel):
    """Schema for single detection request."""

    content: str = Field(..., min_length=1, max_length=5000)
    include_analysis: bool = True
    include_propagation: bool = False


class DetectionResponse(BaseModel):
    """Schema for detection response."""

    id: uuid.UUID
    content: str
    is_rumor: bool
    confidence: float = Field(..., ge=0.0, le=1.0)
    risk_level: RiskLevel
    explanation: str
    analysis: Optional[AnalysisResult] = None
    created_at: datetime

    model_config = {"from_attributes": True}


class BatchDetectionRequest(BaseModel):
    """Schema for batch detection request."""

    contents: list[str] = Field(..., min_length=1, max_length=100)
    include_analysis: bool = True


class BatchDetectionResponse(BaseModel):
    """Schema for batch detection response."""

    total: int
    success: int
    failed: int
    results: list[DetectionResponse]


class PropagationNode(BaseModel):
    """Schema for propagation node in network visualization."""

    node_id: str
    parent_id: Optional[str] = None
    content: Optional[str] = None
    user_info: Optional[dict] = None
    engagement: Optional[dict] = None
    timestamp: Optional[datetime] = None


class PropagationResponse(BaseModel):
    """Schema for propagation analysis response."""

    detection_id: uuid.UUID
    nodes: list[PropagationNode]
    pattern: Optional[str] = None
    spread_speed: Optional[str] = None
    estimated_reach: Optional[int] = None
    influence_score: Optional[float] = None
